crates/aptos-logger/src/speculative_log.rs

```diff
@@ -0,0 +1,284 @@
+// Copyright (c) Aptos
+// SPDX-License-Identifier: Apache-2.0
+
+use crate as aptos_logger;
+/// Infrastucture for logging speculatively, i.e. ability to clear prior logs.
+use crate::{debug, error, info, trace, warn, Level};
+use aptos_crypto::HashValue;
+use aptos_infallible::Mutex;
+use aptos_log_derive::Schema;
+use aptos_state_view::StateViewId;
+use aptos_types::transaction::Version;
+use arc_swap::ArcSwap;
+use crossbeam::utils::CachePadded;
+use once_cell::sync::Lazy;
+use rayon::prelude::*;
+use serde::Serialize;
+use std::sync::{
+    atomic::{AtomicUsize, Ordering},
+    Arc,
+};
+
+#[derive(Default)]
+struct BufferedLog {
+    entries: Vec<(Level, AdapterLogSchema, String)>, // Log level & the message string
+    critical_err_cnt: usize,
```

Comment:
What is the difference between `critical_err_cnt` here and the `AtomicUsize` inside `BUFFERED_LOGS`? Adding some comments for clarity would be great

Reply:
Good point, will add comment. the outer one is sum, so when we flush, we don't have to do a traversal.